### PR TITLE
RSPEC-1764 Java naive implementation for Binary operators identical expressions

### LINKED ISSUE #37: Rule S1764: Identical expressions should not be used on both sides of a binary operator
Implements [RSPEC-1764](https://jira.sonarsource.com/browse/RSPEC-1764)

### PR DESCRIPTION
**Pull request by ghost**
Fixes  #37

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+  }
+
+  private static boolean isExcluded(BinaryExpressionLike binaryExpression) {
+    String operator = binaryExpression.operatorToken.value;
+    // "=" not considered as binary operator for Java and Go, might be for other languages?
+    return "+".equals(operator) || "*".equals(operator) || "=".equals(operator);
+  }
+
+  private static class BinaryExpressionLike {
```
Comment (saberduck) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java]: I think this class should be in the package `uast.helpers` like other `*Like` classes

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/CheckList.java [hunk 2]
```diff
@@ -10,7 +10,8 @@
       NoIdenticalFunctionsCheck.class,
       NoHardcodedCredentialsCheck.class,
       NoSelfAssignmentCheck.class,
-      TooManyParametersCheck.class
+      TooManyParametersCheck.class,
+      BinaryOperatorIdenticalExpressionsCheck.class
```
Comment (saberduck) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/CheckList.java]: Let's keep this ordered alphabetically

File: uast-generator-go/goparser.go [hunk 3]
```diff
@@ -313,7 +314,7 @@ func mapExpr(astNode ast.Expr) *Node {
 
 func mapBinaryExpr(expr *ast.BinaryExpr) *Node {
 	return &Node{
-		Kinds: kinds(expr),
+		Kinds: kinds(BINARY_EXPRESSION),
```
Comment (saberduck) [inline uast-generator-go/goparser.go]: I think it's more consistent to add case for `*ast.BinaryExpr` to the switch in `kind` func , than calling `kinds` with explicit `kind`. @janos-ss can you please confirm?

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java (hunk 1 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java]: Also this should be explicitly unit tested

**Review by saberduck — APPROVED** [review]
Please consider my comments, otherwise LGTM

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/CheckList.java (hunk 2 above)
Comment (ghost) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/CheckList.java]: Didn't noticed it before

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java (hunk 1 above)
Comment (ghost) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/BinaryOperatorIdenticalExpressionsCheck.java]: Ok, I was wondering about that. But let's make it common.

File: uast-generator-go/goparser.go [hunk 4]
```diff
@@ -313,7 +316,7 @@ func mapExpr(astNode ast.Expr) *Node {
 
 func mapBinaryExpr(expr *ast.BinaryExpr) *Node {
 	return &Node{
-		Kinds: kinds(expr),
+		Kinds: kinds(kind(expr)),
```
Comment (ghost) [inline uast-generator-go/goparser.go]: You can write as `kinds(expr)`

**Review by ghost — APPROVED** [review]

File: uast-generator-go/goparser_test.go [hunk 5]
```diff
@@ -275,6 +275,48 @@ func Test_mapIfStmt(t *testing.T) {
...
+	v16 := 1 << 2
+	v17 := 1 >> 2
+	v18 := 1 & 2
+	v19 := 1 &^ 2
+}
+`
+	expectedOperators := []string{"||", "&&", "==", "!=", "<", "<=", ">", ">=", "+", "-", "|", "^", "*", "/", "%", "<<", ">>", "&", "&^"}
+	fileSet, astFile := astFromString(source)
+	blockStmt := astFile.Decls[0].(*ast.FuncDecl).Body
+	for i := 0; i < len(blockStmt.List); i++ {
```
Comment (ghost) [inline uast-generator-go/goparser_test.go]: I think a `range` loop would be more idiomatic here: `for i, stmt := range blockStmt.List {`

**Review by ghost — APPROVED** [review]

**Pull request merged**